File: Documentation/3.0/Samples/csharp/Raven.Documentation.Samples/ClientApi/Changes/HowToSubscribeToDataSubscriptionChanges.cs
using System;

using Raven.Abstractions.Data;
using Raven.Client.Changes;
using Raven.Client.Document;

namespace Raven.Documentation.Samples.ClientApi.Changes
{
	public class HowToSubscribeToDataSubscriptionChanges
	{
		private interface IFoo
		{
			#region data_subscription_changes_1
			IObservableWithTask<DataSubscriptionChangeNotification>
				ForAllDataSubscriptions();
			#endregion

			#region data_subscription_changes_3
			IObservableWithTask<DataSubscriptionChangeNotification> 
				ForDataSubscription(long id);
			#endregion
		}

		public HowToSubscribeToDataSubscriptionChanges()
		{
			using (var store = new DocumentStore())
			{
				{
					#region data_subscription_changes_2

					IDisposable subscription = store
						.Changes()
						.ForAllDataSubscriptions()
						.Subscribe(
							change =>
							{
								var subscriptionId = change.Id;

								switch (change.Type)
								{
									case DataSubscriptionChangeTypes.SubscriptionOpened:
										// do something
										break;
									case DataSubscriptionChangeTypes.SubscriptionReleased:
										// do something
										break;
								}
							});

					#endregion
				}
				{
					#region data_subscription_changes_4

					var subscriptionId = 3;

					IDisposable subscription = store
						.Changes()
						.ForDataSubscription(subscriptionId)
						.Subscribe(
							change =>
							{
								switch (change.Type)
								{
									case DataSubscriptionChangeTypes.SubscriptionOpened:
										// do something
										break;
									case DataSubscriptionChangeTypes.SubscriptionReleased:
										// do something
										break;
								}
							});

					#endregion
				}
			}
		}
	}
}

File: Documentation/3.5/Samples/java/src/test/java/net/ravendb/clientapi/changes/HowToSubscribeToDataSubscriptionChanges.java
package net.ravendb.clientapi.changes;

import net.ravendb.abstractions.basic.CleanCloseable;
import net.ravendb.abstractions.closure.Action1;
import net.ravendb.abstractions.data.DataSubscriptionChangeNotification;
import net.ravendb.client.changes.IObservable;
import net.ravendb.client.document.DocumentStore;
import net.ravendb.client.utils.Observers;

public class HowToSubscribeToDataSubscriptionChanges {


    @SuppressWarnings("unused")
    private interface IFoo {

        //region data_subscription_changes_1
        IObservable<DataSubscriptionChangeNotification> forAllDataSubscriptions();
        //endregion

        //region data_subscription_changes_3
        IObservable<DataSubscriptionChangeNotification> forDataSubscription(long id);
        //endregion
    }

    public HowToSubscribeToDataSubscriptionChanges() {
        try (DocumentStore store = new DocumentStore()) {
            {
                //region data_subscription_changes_2
                CleanCloseable subscription = store.
                        changes()
                        .forAllDataSubscriptions()
                        .subscribe(Observers.create(new Action1<DataSubscriptionChangeNotification>() {
                            @Override
                            public void apply(DataSubscriptionChangeNotification change) {
                                long subscriptionId = change.getId();

                                switch (change.getType()) {
                                    case SUBSCRIPTION_OPENED:
                                        // do something
                                        break;
                                    case SUBSCRIPTION_RELEASED:
                                        // do something
                                        break;
                                }
                            }
                        }));
                //endregion

                //region data_subscription_changes_4
                int subscriptionId = 3;

                CleanCloseable subscription2 = store.
                        changes()
                        .forDataSubscription(3)
                        .subscribe(Observers.create(new Action1<DataSubscriptionChangeNotification>() {
                            @Override
                            public void apply(DataSubscriptionChangeNotification change) {
                                long subscriptionId = change.getId();

                                switch (change.getType()) {
                                    case SUBSCRIPTION_OPENED:
                                        // do something
                                        break;
                                    case SUBSCRIPTION_RELEASED:
                                        // do something
                                        break;
                                }
                            }
                        }));

                //endregion
            }
        }
    }
}
